### PR TITLE
[IMP] Check if the folder has changed and then build it

### PR DESCRIPTION
**Pull request by JesusZapata**
Referring this comment https://github.com/Vauxoo/docker-odoo-image/pull/191#issuecomment-278669714 and this TODO in .travis.yml
` # TODO: Check git diff changes to build just required image.`

### CODE REVIEW

**Comment by JesusZapata** [comment]
@moylop260 @ruiztulio 
What do you think of this PR?
Only build the images that were changed

File: build.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    def _has_been_changed(self):
+        return self._dircmp(filecmp.dircmp(self._folder,
+                                           os.path.join(self._tempdir,
+                                                        self._relpath)))
+
+    def build(self):
+        can_be_build = True
+        if self._git_check:
+            if not os.path.isdir(self._tempdir):
+                cmd = ["git", "clone",
```
Comment (moylop260) [inline build.py]: You don't need `clone` to check if a file has been changed.
You could use directly the remote like as [MQT-git_run](https://github.com/Vauxoo/maintainer-quality-tools/blob/233d67accf3da40f628ab64c49164ff6d9c9c6cb/travis/git_run.py#L24)

Comment (moylop260) [inline build.py]: FYI you have the following environment variables to work:
https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

Notice `TRAVIS_PULL_REQUEST`, `TRAVIS_BRANCH` and `TRAVIS_BUILD_DIR`

Comment (JesusZapata) [inline build.py]: @moylop260 
Thanks for the information about `git diff`

**Comment by JesusZapata** [comment]
@moylop260 @ruiztulio 
Now you are only going to build the image that was modified
This will allow a shorter wait time for developers

**Comment by ruiztulio** [comment]
@JesusZapata LGTM, but is the time required to download the ubuntu images lower than the time required to build the image?

**Comment by JesusZapata** [comment]
@ruiztulio 

The time in `git clone https://github.com/vauxoo/docker-ubuntu-base ${TRAVIS_BUILD_DIR}/docker-ubuntu-base` is **1.72 s** https://travis-ci.org/Vauxoo/docker-odoo-image/builds/201121217#L257

The time in `python ${TRAVIS_BUILD_DIR}/build.py --folder ${TRAVIS_BUILD_DIR}/docker-ubuntu-base --docker-image vauxoo/docker-ubuntu-base:latest --no-git-check` **240.75 s** https://travis-ci.org/Vauxoo/docker-odoo-image/builds/201121217#L275

The build Is greater than the clone

Does that answer your question?

**Comment by ruiztulio** [comment]
@JesusZapata Sorry, I meant the time to download the image vauxoo/docker-ubuntu-base from hub.docker.com, the actual .travis [build the ubuntu images](https://github.com/Vauxoo/docker-odoo-image/blob/master/.travis.yml#L13-L14) every time we do a test and (if I understood well) with your changes it won't, it will download the image from [docker hub](https://hub.docker.com/r/vauxoo/docker-ubuntu-base/) to shorten the total build time (and shorten the time by not building the odoo-100 and odoo-80 if only a change was made in odoo-shippable)

So my question originally was, the build time is longer than the download time? 

And as suggestion consider not to build/download the ubuntu images if there was no change in odoo-80 and odoo-100 (odoo-80 uses ubuntu:latest and odoo-100 uses ubuntu:16.04) this will shorten even more the total time because for a change in odoo-shippable we only need the odoo-80 image foe example

**Comment by JesusZapata** [comment]
@ruiztulio 
Docker pull is most faster and clearer than the clone and build
And now the .travis.yml is very simple

**Comment by ruiztulio** [comment]
@JesusZapata where is the information you told me in the chat?

a

**Comment by JesusZapata** [comment]
@ruiztulio 

The build of odoo80 `python ${TRAVIS_BUILD_DIR}/build.py --folder ${TRAVIS_BUILD_DIR}/odoo80 --docker-image vauxoo/odoo-80-image:latest` It takes **784 s** https://travis-ci.org/Vauxoo/docker-odoo-image/builds/201251742#L257

The build of `cd ${TRAVIS_BUILD_DIR}/docker-ubuntu-base && docker build --rm -t vauxoo/docker-ubuntu-base:latest .` It takes **183 s** https://travis-ci.org/Vauxoo/docker-odoo-image/builds/194393642#L393
And the build of `cd ${TRAVIS_BUILD_DIR}/odoo80 && docker build --rm -t vauxoo/odoo-80-image:latest .` It takes **765 s** https://travis-ci.org/Vauxoo/docker-odoo-image/builds/194393642#L2554

The current version with docker pull it takes **784 s**
The previus version with git pull and docker build it takes **948 s**

This represents a savings of **164 s**

**Comment by moylop260** [comment]
Could you squash your commits all-in-one, please?

**Comment by JesusZapata** [comment]
@moylop260 
All commit are in one

File: build.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import argparse
+
+
+class DockerOdooImages(object):
+
+    def __init__(self, folder, docker_image):
+        self._folder = folder
+        self._relpath = os.path.relpath(self._folder)
+        self._docker_file_path = os.path.join(self._folder, 'Dockerfile')
+        self._docker_file = open(self._docker_file_path).read().split('\n')
```
Comment (moylop260) [inline build.py]: Remove unused `self._docker_file` variable

**Comment by moylop260** [comment]
Run a `flake8` and `pylint` to fix a few checks please

**Comment by JesusZapata** [comment]
@moylop260 
Now flake8 and pep8 are checked

image

**Comment by ruiztulio** [comment]
@moylop260 @JesusZapata LGTM

**Pull request merged**